src/storage/secondary/RedisStorage.cpp

```diff
@@ -0,0 +1,353 @@
+  if (si == suffix.begin()) {
+    if (suffix[0] == '/') {
+      sock = suffix;
+    } else {
+      host = suffix;
+    }
+  } else {
+    host.assign(suffix.begin(), si - 1);
+    port.assign(si, suffix.end());
+  }
+  if (m_connect_timeout) {
+    LOG("Redis connect timeout {}", timeval_to_string(*m_connect_timeout));
+  }
+  if (!host.empty()) {
+    int p = port.empty() ? 6379 : std::stoi(port);
+    if (m_connect_timeout) {
+      m_context = redisConnectWithTimeout(host.c_str(), p, *m_connect_timeout);
+    } else {
+      m_context = redisConnect(host.c_str(), p);
+    }
+  } else if (!sock.empty()) {
+    if (m_connect_timeout) {
+      m_context = redisConnectUnixWithTimeout(sock.c_str(), *m_connect_timeout);
+    } else {
+      m_context = redisConnectUnix(sock.c_str());
+    }
+  } else {
+    LOG("Redis invalid url: {}", m_url);
+    m_invalid = true;
+    return false;
+  }
+
+  if (!m_context || m_context->err) {
+    LOG("Redis connect {} err: {}", m_url, m_context->errstr);
+    m_invalid = true;
+    return false;
+  } else {
+    if (m_context->connection_type == REDIS_CONN_TCP) {
+      LOG(
+        "Redis connect tcp {}:{} OK", m_context->tcp.host, m_context->tcp.port);
+    }
+    if (m_context->connection_type == REDIS_CONN_UNIX) {
+      LOG("Redis connect unix {} OK", m_context->unix_sock.path);
+    }
+    m_connected = true;
+
+    if (m_operation_timeout) {
+      LOG("Redis timeout {}", timeval_to_string(*m_operation_timeout));
+      if (redisSetTimeout(m_context, *m_operation_timeout) != REDIS_OK) {
+        LOG_RAW("Failed to set timeout");
+      }
+    }
+
+    if (m_password) {
+      std::string username = m_username ? *m_username : "default";
+      LOG("Redis AUTH {} {}", username, "*****"); // don't log m_password !!!
+      redisReply* reply;
+      if (m_username) {
+        reply = static_cast<redisReply*>(redisCommand(
+          m_context, "AUTH %s %s", m_username->c_str(), m_password->c_str()));
+      } else {
+        reply = static_cast<redisReply*>(
+          redisCommand(m_context, "AUTH %s", m_password->c_str()));
+      }
+      if (!reply) {
+        LOG("Failed to auth {} in redis", username);
+        m_invalid = true;
+      } else if (reply->type == REDIS_REPLY_ERROR) {
+        LOG("Failed to auth {} in redis: {}", username, reply->str);
+        m_invalid = true;
+      }
+      freeReplyObject(reply);
+      if (m_invalid) {
+        return false;
+      }
+    }
+
+    return true;
+  }
+}
+
+void
+RedisStorage::disconnect()
+{
+  if (m_connected) {
+    // Note: only the async API actually disconnects from the server
+    //       the connection is eventually cleaned up in redisFree()
+    LOG_RAW("Redis disconnect");
+    m_connected = false;
+  }
+}
+
+nonstd::expected<nonstd::optional<std::string>, SecondaryStorage::Error>
+RedisStorage::get(const Digest& digest)
+{
+  if (!connect()) {
+    return nonstd::make_unexpected(Error::error);
+  }
```

Comment:
This is exactly what I had in mind, thanks! :+1:

However, a heads-up: when working on statistics for secondary storage, I've realized that I need to remember which backends are in an error state anyway to be able to only increment "secondary storage error" once. Therefore I plan to instead let the framework create a backend only when needed and remove it from the list of backends after the first error/timeout. This should also simplify backends so that you can connect in the constructor and don't have to remember the error state yourself.